File: ReorderList.java

public class Solution {
    public void reorderList(ListNode head) {
        if (head == null || head.next == null)
            return;
        ListNode quick = head, slow = head;
        while(quick != null && quick.next != null && quick.next.next != null) {
            quick = quick.next.next;
            slow = slow.next;
        }
        ListNode second_half = slow.next;
        slow.next = null;
        printList(second_half);

        second_half = reverseList(second_half);
        printList(head);
        printList(second_half);
        ListNode first_half = head;
        while(second_half != null) {
            ListNode temp = second_half.next;
            second_half.next = first_half.next;
            first_half.next = second_half;
            first_half = second_half.next;
            second_half = temp;
        }

    }

    public ListNode reverseList(ListNode head) {
        if(head == null || head.next == null)
            return head;
        ListNode previous = null;
        ListNode current = head;
        ListNode current_next = head.next;
        while(current_next != null) {
            current.next = previous;
            previous = current_next;
            ListNode temp = current_next.next;
            current_next.next = current;
            current = temp;
            if (current != null)
                current_next = current.next;
            else
                break;
        }
        if (current != null && current.next == null) {
            current.next = previous;
            return current;
        }
        return current_next;
    }

    public void printList(ListNode node) {
        while(node != null) {
            System.out.print(node.val + " -> ");
            node = node.next;
        }
        System.out.println("NULL");
    }

    public static void main (String [] args) {
        Solution mysolution = new Solution();
        ListNode first = new ListNode(1);
        ListNode second = new ListNode(2);
        ListNode third = new ListNode(3);
        ListNode fourth = new ListNode(4);
        /*
        ListNode fifth = new ListNode(5);
        ListNode sixth = new ListNode(6);
        ListNode seventh = new ListNode(7);
        */
        first.next = second;
        second.next = third;
        third.next = fourth;
        /*
        fourth.next = fifth;
        fifth.next = sixth;
        sixth.next = seventh;
        */
        fourth.next = null;
        mysolution.printList(first);
        mysolution.reorderList(first);
        mysolution.printList(first);
    }
}

File: ReorderList.cc
/**
* Definition for singly-linked list.
* struct ListNode {
*     int val;
*     ListNode *next;
*     ListNode(int x) : val(x), next(NULL) {}
* };
*/
class Solution {
public:
    void reorderList(ListNode *head) {
        if (!head || !head->next) {
            return;
        }
        int length = 0;
        ListNode *current = head;
        while(current) {
            length++;
            current = current->next;
        }
        if (length < 3) {
            return;
        }
        current = head;
        int half = length / 2;
        while(--half > 0) {
            current = current->next;
        }
        if (length % 2 == 1) {
            current = current->next;
        }
        ListNode *second_half = reverseList_2(current->next);
        if (!second_half) {
            return;
        }
        /* tangle them */
        current = head;
        ListNode *first_half_next = head->next;
        ListNode *second_half_next = second_half->next;
        while(second_half) {
            current->next = second_half;
            second_half->next = first_half_next;
            second_half = second_half_next;
            if (second_half_next) {
                second_half_next = second_half_next->next;
            }
            current = first_half_next;
            first_half_next = first_half_next->next;
        }
        current->next = NULL;
    }

    ListNode *reverseList(ListNode *node) {
        if (!node || !node->next) {
            return node;
        }

        ListNode *newhead = reverseList(node->next);
        node->next->next = node;
        node->next = NULL;
        return newhead;
    }

    ListNode * reverseList_2(ListNode *node) {
        if (!node) {
            return NULL;
        }

        if (!node->next) {
            return node;
        }

        ListNode *current = node;
        ListNode *prev = NULL;
        while(current) {
            ListNode *temp = current->next;
            current->next = prev;
            prev = current;
            current = temp;
        }

        return prev;
    }
};
